akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -472,13 +472,17 @@ final class Split[T](val decision: Split.SplitDecision, val p: T ⇒ Boolean, va
         setKeepGoing(enabled = handler.hasInitialElement)
 
         if (isAvailable(out)) {
-          push(out, Source.fromGraph(substreamSource.source))
-          scheduleOnce(SubscriptionTimer, timeout)
-          substreamWaitingToBePushed = true
-        } else substreamWaitingToBePushed = false
+          if (decision == SplitBefore || handler.hasInitialElement) pushSubstreamSource() else pull(in)
+        } else substreamWaitingToBePushed = true
       }
     }
 
+    private def pushSubstreamSource(): Unit = {
+      push(out, Source.fromGraph(substreamSource.source))
+      scheduleOnce(SubscriptionTimer, timeout)
+      substreamWaitingToBePushed = false
+    }
```

Comment:
And then also `substreamWaitingToBeStarted` or something corresponding to whatever this method is named